community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala

```diff
@@ -52,29 +54,60 @@ case class VarLengthExpandPipe(source: Pipe,
                                min: Int,
                                max: Option[Int],
                                nodeInScope: Boolean,
-                               filteringStep: VarlenghtPredicate = VarlenghtPredicate.NONE)
+                               filteringStep: VarLengthPredicate = VarLengthPredicate.NONE)
                               (val estimatedCardinality: Option[Double] = None)
                               (implicit pipeMonitor: PipeMonitor) extends PipeWithSource(source, pipeMonitor) with RonjaPipe {
 
+  trait ContainsRelationship {
+    def contains(r: Relationship): Boolean
+  }
+
+  class SetBacked(set: mutable.Set[Relationship]) extends ContainsRelationship {
+    override def contains(r: Relationship): Boolean = set.contains(r)
+  }
+
+  class VectorBacked(set: Vector[Relationship]) extends ContainsRelationship {
+    override def contains(r: Relationship): Boolean = set.contains(r)
+  }
+
+  private val needsFlipping = if (dir == SemanticDirection.BOTH)
+    projectedDir == SemanticDirection.INCOMING
+  else
+    dir != projectedDir
+
   private def varLengthExpand(node: Node, state: QueryState, maxDepth: Option[Int],
                               row: ExecutionContext): Iterator[(Node, Seq[Relationship])] = {
-    val stack = new mutable.Stack[(Node, Seq[Relationship])]
-    stack.push((node, Seq.empty))
-
-    new Iterator[(Node, Seq[Relationship])] {
-      def next(): (Node, Seq[Relationship]) = {
-        val (node, rels) = stack.pop()
-        if (rels.length < maxDepth.getOrElse(Int.MaxValue) && filteringStep.filterNode(row,state)(node)) {
-          val relationships: Iterator[Relationship] = state.query.getRelationshipsForIds(node, dir, types.types(state.query))
+    val maxDepthValue = maxDepth.getOrElse(Int.MaxValue)
+
+    val nodeFilter: (Node) => Boolean = filteringStep.filterNode(row, state)
+
+    val stack = new mutable.ArrayBuffer[(Node, Vector[Relationship])](1024 * 1024)
+    stack.append((node, Vector[Relationship]()))
+    val relSet = mutable.HashSet[Relationship]()
+
+    new Iterator[(Node, Vector[Relationship])] {
+      override def next(): (Node, Vector[Relationship]) = {
+        val (node, rels: Vector[Relationship]) = stack.remove(stack.length - 1)
+        if (rels.length < maxDepthValue && nodeFilter(node)) {
+          val relationships: Iterator[Relationship] = state.query
+            .getRelationshipsForIds(node, dir, types.types(state.query))
+            .filter(filteringStep.filterRelationship(row, state))
+
+          val set: ContainsRelationship = if (rels.length > HASH_SET_CUTOFF) {
```

Comment:
You should at some point have a look how the `Vector.contains()` works internally, it's really bad complexity :( and shows up as 90% cpu in profiling.